### PR TITLE
Add Data Plane Docs

### LINKED ISSUE #7: Data Plane Specification - Initial Thoughts
Some early thoughts on the data plane layer to garner feedback on directions. 

## Goals

 * Provide a data payload that a range of predefined ML Servers including TFServing, XGBoost, SkLearn can interface to handle prediction requests and return responses to an external client.
 * Allow custom predict functions to be built that can serve requests using the API.
 
## Initial Design Questions

  1. Should we provide a new data plane spec or allow the control plane components to specify which data plane they will handle?

The reasons to not provide an initial data plane spec are:

 * We want to provide flexibility.
 * We want to utilize initial payloads specifications which will allow for speedy implementation using existing components such as Seldon's Prediction payload or TFServing payload.
 * We expect data science models to cover a wide variety of use cases so its unlikely a single spec can be both generic enough and simple enough.

For the above the control plane specs would be extended with a `protocol` field, e.g.

```
apiVersion: "serving.kubeflow.org/v1alpha1"
kind: "KFService"
metadata:
  name: "myModel"
  spec:
    minReplicas: 1
    maxReplicas: 2 
    tensorflow:
      modelUri: "gs://mybucket/mymodel"
      protocol: seldon # could also be tfserving
```

### Payloads

There are a range of payload types that can be provided each with their own advantages.

 * TensorProto
   * Standard for Tensorflow Serving
 * NDArray 
   * Useful for multi-typed data and REST JSON transport
 * Simplified Tensor (shape and array of doubles)
   * Easier for REST JSON use by beginners

### Meta Data

Should we provide a place to provide meta data as part of the payload? Meta data in the API could include:

 * Unique Prediction ID
 * Arbitrary map of key/value tags
 * Meta data about the model that is serving the request
 * Metrics created

### Status

Should there be a section of the payload providing status of the prediction?

 * Success/Failure
 * Error codes and human readable descritpions

## Support for CloudEvents

If we wish to support multiple payload protocols and at the same time handle synchronous and asynchronous use cases we may wish to contain the payload within a meta-API that defines what schema the underlying payload contains.

[Cloudevents](https://cloudevents.io/) provides a generic protocol for events. By supporting cloudevents as the top level protocol we can easily create data specifications for various ML use cases at that level rather than in 1 overarching data plane definition:

  * Prediction (existing concern for current discussion)
  * Reinforcement learning
    * Send feedback on previous predictions as to whether they were correct
    * Provide current world state and get new action

In this way we can build up a set of data plane protocols for different use cases and/or handle particular exisiting protocols (Seldon, TFServing). CloudEvents also fits well into the KNative ecosystem.

Example:

```
{
      "specversion" : "0.2",
      "type" : "org.mlspec.serving.prediction",
      "schemaurl": "http://mlspec.org/serving/prediction.proto"
      "source" : "/myclient",
      "id" : "C234-1234-1234",
      "time" : "2018-04-05T17:31:05Z",
      "datacontenttype" : "application/json",
      "data" : {
         "meta" : { "hyperParam1":"value1" },
         "tensor": {
	  "values": [1.0,0.5,0.2],
	  "shape": [1,3]
	 }
      }
}
```

a response could be a cloudevent:

```
{
      "specversion" : "0.2",
      "type" : "org.mlspec.serving.prediction",
      "schemaurl": "http://mlspec.org/serving/prediction.proto"
      "source" : "/mymodel1",
      "id" : "C234-1234-1234",
      "time" : "2018-04-05T17:31:05Z",
      "datacontenttype" : "application/json",
      "data" : {
         "meta" : {
	   "model":"resetNetv1.1",
           "status":"success"
          },
         "tensor": {
	  "values": [0.9],
	  "shape": [1]
	 }
      }
}
```

This may be useful in future pipeline use cases where components of the pipeline may only be interested in some subset of events. So all components may respond and transform `prediction` events but only some to reinforcement learning `feedback` events.

### PR DESCRIPTION
**Pull request by cliveseldon**
* Adds an initial data plane doc from current discussions
  * Refactors README to point to control plane and data plane docs

Fixes #7 

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/32)

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -1,155 +1,8 @@
...
+# Design Docs
 
-| Field       |  Value      | Description |
-| ----------- | ----------- | ----------- |
-| kind       | KFService                     | [Read the Docs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#types-kinds) |
-| apiVersion | serving.kubeflow.org/v1alpha1 | [Read the Docs](https://kubernetes.io/docs/reference/using-api/api-overview/#api-versioning) |
-| metadata   | [Metadata](#Metadata)         | [Read the Docs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata) |
-| spec       | [Spec](#Spec)                 | [Read the Docs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status) |
-| status     | [Status](#Status)             | [Read the Docs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status) |
+ * [Kfserving control plane specification](control-plane.md)
```
Comment (ellis-bigelow) [inline docs/README.md]: nit Kf -> KF

File: docs/README.md [hunk 2]
```diff
@@ -1,155 +1,8 @@
...
+   * Defines the specification of the `kfserving` resource.
+ * [Kfserving data plane specification](data-plane.md)
+   * Defines the data payloads for `kfserving` components.
```
Comment (ellis-bigelow) [inline docs/README.md]:
```suggestion
   * Defines the data payloads to enable interoperability between `kfserving` components.
```

File: docs/data-plane.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
+# Data Plane Specification
+
+This document is a work in progress to disucss the data plane requirements for machine learning serving.
+
+Its aims:
+
+ * Provide a set of schemas for generic machine learning  components, such as model servers, explainers, outlier detectors.
```
Comment (ellis-bigelow) [inline docs/data-plane.md]: nit 
```suggestion
 * Provide a set of schemas for generic machine learning  components, such as predictors, explainers, outlier detectors.
```

File: docs/data-plane.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+## Combined Schema
+
+There is an open question whether we define a combined schema to return the aggregation from the various components or we assume only the model response is returned and other components (model explanation etc) return their response asychronously to some metrics/logging channel.
+
+
+## Metadata
+
+  * There should be a unified prediction id so responses from varied components can be tied together for monitoring and auditing.
+
+It unclear whether we should impose any other metadata.
```
Comment (ellis-bigelow) [inline docs/data-plane.md]: I know this is WIP draft, but consider adding a note about turning on or off features (explain/skew) from the data plane

File: docs/data-plane.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+## Schema Publishing
+
+Components should be able to advertise what schemas they respect to allow the control plane to do static validation. Static validation will be important if we allow pipelines of components in future.
+
+KNative has a [propsoal](https://github.com/knative/eventing/blob/6155ebb1f662e7d8930d27d3446e47103bde5c85/docs/registry/README.md) in the context of KNative eventing.
```
Comment (ellis-bigelow) [inline docs/data-plane.md]:
```suggestion
Knative has a [propsoal](https://github.com/knative/eventing/blob/6155ebb1f662e7d8930d27d3446e47103bde5c85/docs/registry/README.md) in the context of Knative eventing.
```

**Comment by ellis-bigelow** [comment]
Mostly LGTM. A few comments.

File: docs/control-plane.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    traffic: 90
+  canary:
+    ready: true
+    replicas: 1
+    traffic: 10
+```
+
+### Conditions
+Conditions provide realtime feedback to users on the underlying state of their deployment. As the feature set of Service grows, additional Conditions will be added to the set. We will use the convention of positive-true conditions; the resource is healthy if there are no conditions who's status is False. If no conditions are present on the resource (e.g. Ready is not present), the resource is not healthy. Only one condition for each type can be available.  
+
```
Comment (hougangliu) [inline docs/control-plane.md]: s/who's status/whose status

**Comment by cliveseldon** [comment]
Ping @ellis-bigelow - ready to approve?

**Comment by ellis-bigelow** [comment]
Thanks Clive!
/lgtm
/approve

**Pull request merged**